File: Python-For-Everyone-Horstmann/Chapter5-Functions/P5.16.py
# Write a recursive function
#     def isPalindrome(string)
# that returns  True if  string is a palindrome, that is, a word that is the same when
# reversed. Examples of palin dromes are “deed”, “rotor”, or “aibohphobia”. Hint: A
# word is a palindrome if the first and last letters match and the remainder is also a
# palindrome.

# FUNCTIONS
def isPalindrome(string):
    if len(string) <= 1:
        return

    firstLetter = string[0]
    lastLetter = string[-1]

    if firstLetter == lastLetter:
        end = len(string) - 2 + 1
        return isPalindrome(string[1:end])

    else:
        return False

def main():
    string = str(input("Enter a string: "))
    print(string, "is", end = " ")

    if isPalindrome(string) == False:
        print("not", end = " ")

    print("a palindrome")

# PROGRAM RUN
main()

File: Java-for-Everyone-Late-Objects-Horstmann/Chapter-5-Methods/P5_16.java
/*Write a recursive method
    public static boolean isPalindrome(String str)
that returns true if str is a palindrome, that is, a word that is the same when reversed.
Examples of palin­drome are “deed”, “rotor”, or “aibohphobia”. Hint: A word is a
palindrome if the first and last letters match and the remainder is also a palindrome.*/

public class P5_16 {
    public static boolean isPalindrome(String str) {
        if (str.length() <= 1) {
            return true;
        }
        else if (str.charAt(0) == str.charAt(str.length() - 1)) {
            return isPalindrome(str.substring(1, str.length() - 1));
        }
        else {
            return false;
        }
    }
}
